### PR TITLE
FEATURE: Add typescript declaration file output

### PR DESCRIPTION
**Pull request by thebigredgeek**
This PR should add native typescript support, as it outputs declaration files that can be consumed by typescript node apps.

### CODE REVIEW

**Comment by thebigredgeek** [comment]
For now, published here: https://www.npmjs.com/package/@boltsource/ioredis

**Comment by calebboyd** [comment]
@thebigredgeek Is this intended to be a WIP? #1053 is talking about generating the types, while this repo is written in Typescript it is still making heavy use of `any` and dynamic assignments.

**Comment by thebigredgeek** [comment]
No this is not WIP, and I'm already using my fork successfully in a large project.  This simply exports the typedefs for use in typescript projects.

**Comment by zen0wu** [comment]
While this is good, it does not have the commands the the current @types/ioredis package provides like `bitcount`, `strlen`, it seems

**Comment by yovanoc** [comment]
waiting for this..

**Comment by jhurliman** [comment]
Are there any blockers for this?

File: tsconfig.json [hunk 1]
```diff
@@ -7,10 +7,8 @@
     "moduleResolution": "node",
     "types": ["node", "mocha"],
     "typeRoots": ["node_modules/@types"],
-    "module": "commonjs"
+    "module": "commonjs",
+    "declaration": true
```
Comment (gioragutt) [inline tsconfig.json:11]: @thebigredgeek don't you also have to add `"types": "index.d.ts"` to `package.json`?

**Review by micalevisk — APPROVED** [review]

File: tsconfig.json (hunk 1 above)
Comment (CatsMiaow) [reply tsconfig.json:11]: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package

**Comment by luin** [comment]
Sorry, not sure why I didn't reply to this.

I'm working on official declarations in https://github.com/luin/ioredis/pull/1504. The major work is to generate declarations for Redis command methods as most of them are generated dynamically, and without handling this the declaration is not useful.

**Comment by luin** [comment]
ioredis v5 provides official TypeScript declarations so closing this. Sorry for the delay!

**Pull request closed without merging**